### PR TITLE
New feat: + / * for Loss

### PR DESCRIPTION
**Pull request by LucasBoTang**
In general, I overload `+` and `*` operators for `AggregateLoss`.

#### `AggregateLoss` Aggregation:

- Add `+`: Enables aggregation of two `AggregateLoss` instances.
- Add `*`: Adjusts the weight of both `Objective` and `Constraints` within `AggregateLoss`.

#### `Objective` Weight Adjustment:

- Add `*`: Modify the weight of the `Objective`.

#### `Constraint` Weight Modification:

- Updated `*`: Calculate the weight of `Constraints` as `weight * self.weight` instead of just `weight`

### CODE REVIEW

**Comment by LucasBoTang** [comment]
Here is an example of adding two losses that share the same parameter `p` and even use two different types of loss ( `PenaltyLoss` and `BarrierLoss`):

```python 
import torch

from neuromancer.constraint import variable
from neuromancer.loss import PenaltyLoss, BarrierLoss

# parameters
p = variable("p")
p1, p2 = p[:, 0], p[:, 1]

# problem 1:
# variables
x = variable("x")
x1, x2 = x[:, 0], x[:, 1]
# objective function
f = x1 ** 2 + x2 ** 2
obj = [f.minimize(weight=1.0, name="obj")]
# constraints
Q_con = 100
con_1 = Q_con * (x1 + x2 - p1 >= 0)
con_1.name = "c1"
con_2 = Q_con * (x1 + x2 - p1 <= 5)
con_2.name = "c2"
con_3 = Q_con * (x1 - x2 + p2 <= 5)
con_3.name = "c3"
con_4 = Q_con * (x1 - x2 + p2 >= 0)
con_4.name = "c4"
cons = [con_1, con_2, con_3, con_4]
# barrier loss
loss_1 = BarrierLoss(obj, cons)

# problem 2
y = variable("y")
y1, y2 = y[:, 0], y[:, 1]
# objective function
f = (1 - y1) ** 2 + p1 * (y2 - y1 ** 2) ** 2
obj = [f.minimize(weight=1.0, name="obj")]
# constraints
Q_con = 100
con_1 = Q_con * (y1 >= y2)
con_1.name = "c1"
con_2 = Q_con * (y1 ** 2 + y2 ** 2 >= p2 / 2)
con_2.name = "c2"
con_3 = Q_con * (y1 ** 2 + y2 ** 2 <= p2)
con_3.name = "c3"
cons = [con_1, con_2, con_3]
# penalty loss
loss_2 = PenaltyLoss(obj, cons)

# aggregate loss
loss = 0.1 * loss_1 + 1.0 * loss_2

# test
datapoints = {"p": torch.FloatTensor(1, 2).uniform_(0.5, 1.2),
              "x": torch.FloatTensor(1, 2).uniform_(0.0, 3.0),
              "y": torch.FloatTensor(1, 2).uniform_(0.0, 1.0),
              "name": "test"}
print(loss(datapoints)["loss"])
```

It gives us the flexibility to aggregate multiple losses.

**Comment by madelynshapiro** [comment]
Thanks for submitting a PR @LucasBoTang ! Please add some basic unit tests for this new feature and request review from myself and @drgona when your changes are complete.

**Comment by LucasBoTang** [comment]
Sure!

**Comment by LucasBoTang** [comment]
Hi @madelynshapiro ,
I added unit test for `+` and `*` operators within the `AggregateLoss`.

**Comment by LucasBoTang** [comment]
While reviewing the code, I noticed that the barrier term for `BarrierLoss` isn't weighted at [neuromancer/loss.py#L211](https://github.com/pnnl/neuromancer/blob/master/src/neuromancer/loss.py#L211), which causes an inconsistency in the loss output when using `*`. I am not sure if this is an intended of  `BarrierLoss`. However, in practice (e.g. Central Path) we also need the weight of barrier, and it is useful to avoid unbalanced scale between objective function and barrier term.

**Comment by drgona** [comment]
> While reviewing the code, I noticed that the barrier term for `BarrierLoss` isn't weighted at [neuromancer/loss.py#L211](https://github.com/pnnl/neuromancer/blob/master/src/neuromancer/loss.py#L211), which causes an inconsistency in the loss output when using `*`. I am not sure if this is an intended feature of `BarrierLoss`. However, in practice (e.g. Central Path) we also need the weight of barrier, and it is useful to avoid unbalanced scale between objective function and barrier term.

Yes we should include weighting term in the barrier loss as well. This is a bug.

**Comment by LucasBoTang** [comment]
> Yes we should include the weighting term in the barrier loss as well. This is a bug. Feel free to make the change in this PR.

Hi @drgona and @madelynshapiro , I fixed the weight of the barrier in the PR and you can review the code and unit test.

**Comment by madelynshapiro** [comment]
Thanks @LucasBoTang ! I will review this today.

File: src/neuromancer/loss.py [hunk 1]
```diff
@@ -12,11 +12,13 @@
 
 import math
 from abc import ABC, abstractmethod
+import copy
 
 import torch
 import torch.nn as nn
 import numpy as np
 import torch.nn.functional as F
+from torch.nn import Module
```
Comment (madelynshapiro) [inline src/neuromancer/loss.py:21]: unused import - remove

**Review by madelynshapiro — APPROVED** [review]
See review comment on unused import - other than that, looks good! Ready to merge once unused import is removed.

**Review by madelynshapiro — COMMENTED** [review]
See review comment on unused import - other than that, looks good! Ready to merge once unused import is removed.

File: src/neuromancer/loss.py (hunk 1 above)
Comment (madelynshapiro) [reply src/neuromancer/loss.py:21]:
```suggestion
```

**Comment by LucasBoTang** [comment]
Removed

**Comment by madelynshapiro** [comment]
@LucasBoTang Merged into the `pnnl/develop` branch! This is a new branch where we will add core codebase changes in between releases in order to track versioning more carefully. I will let you know when this feature is released to the main branch!
